File: c#/_95_AmicableChains.cs
using System;
using System.Collections.Generic;

public class _95_AmicableChains{
	public static int Euler95(){
		int max = 1_000_000, maxLen = 0, minNum = 0;
		for (int i = 12496; i < max; i++){
			SortedSet<int> chain = new SortedSet<int>();
			int n = i;
			while (!chain.Contains(n) && n < max){
				chain.Add(n);
				n = GetProperDivisorSum(n);
			}
			// found longer chain that returns to starting point
			if (n == i && chain.Count > maxLen){
				// update longest chain length and minimum element
				maxLen = chain.Count;
				minNum = chain.Min;
			}
		}
		return minNum;
    }
	
	public static int GetProperDivisorSum(int n){
		// 1 is always a proper divisor
		// n is a factor but not a proper divisor
		int res = 1;
		for (int i = 2; i*i <= n; i++){
			if (n % i == 0){
				// add pair of factors to proper divisor sum
				res += i + (n / i);
			}
		}
		// avoid double-counting factor if n is a perfect square
		if (Math.Sqrt(n) == (int) Math.Sqrt(n)){
			res -= (int) Math.Sqrt(n);
		}
		return res;
	}
	
	public static void Main(){
		Console.WriteLine(Euler95());
	}
}

File: java/_95_AmicableChains.java
import java.util.*;

public class _95_AmicableChains{
    
    public static int sumProperDivisors(int n){
        int res = 1;
        for (int i = 2; i * i <= n; i++){
            if (n % i == 0){
                res += i + (n / i);
                if (i * i == n) res -= i;
            }
        }
        return res;
    }
    
    public static void main(String[] args){
        int max = 1_000_000, maxLen = 0;
        int res = max;
        for (int n = 1; n < max; n++){
            Set<Integer> chain = new HashSet<>();
            chain.add(n);
            int d = sumProperDivisors(n);
            
            while (!chain.contains(d)){
                chain.add(d);
                int divSum = sumProperDivisors(d);
                if (divSum > max) break;
                d = divSum;
            }
            
            // check if chain returns to starting point
            if (d == n && chain.size() > maxLen){
                // update length of longest chain found and minimum element in longest chain
                maxLen = chain.size();
                res = n;
                for (int num : chain) res = Math.min(res, num);
            }
        }
        
        System.out.println("Smallest member of longest amicable chain with no element exceeding 1 million: " + res); // 14316
    }
}
